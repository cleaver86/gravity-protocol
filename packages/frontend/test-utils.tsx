import React, { ReactElement } from 'react'
import { render, RenderOptions } from '@testing-library/react'
import { ChakraProvider } from '@chakra-ui/react'
import { WalletProvider } from './providers/WalletProvider'
import theme from './theme'

type Providers = {
  children: React.ReactNode
  account?: string
}

const AllTheProviders = ({ children, account }: Providers) => {
  if (!account) {
    throw new Error('AllTheProviders: account must be provided')
  }

  return (
    <ChakraProvider theme={theme}>
      <WalletProvider account={account}>{children}</WalletProvider>
    </ChakraProvider>
  )
}

type Options = Omit<RenderOptions, 'wrapper'> & {
  wrapperProps: {
    account?: string
  }
}

const customRender = (ui: ReactElement, options?: Options) => {
  return render(ui, {
    wrapper: (props) => (
      <AllTheProviders {...props} {...options?.wrapperProps} />
    ),
    ...options,
  })
}

export * from '@testing-library/react'
export { customRender as render }
