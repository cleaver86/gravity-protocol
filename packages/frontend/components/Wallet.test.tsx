import { utils } from 'ethers'
import { useMediaQuery } from '@chakra-ui/react'
import { render, screen, within, waitFor } from '../test-utils'
import userEvent from '@testing-library/user-event'
import {
  WALLET_PROVIDER_METAMASK,
  TOKENS,
  WALLET_PROVIDER_COINBASE,
  WALLET_PROVIDER_WALLETCONNECT,
} from '../constants'
import Wallet from './Wallet'

const MOCK_ACCOUNT_LONG = '0xf61B443A155b07D2b2cAeA2d99715dC84E839EEf'
const MOCK_TOKEN_BALANCE = '1.1234567'
const MOCK_TOKEN_BALANCE_BIG_NUMBER = utils.parseEther(MOCK_TOKEN_BALANCE)
const MOCK_TOKEN_USD_PRICE = 1000.75
const MOCK_TOKEN_ETH_PRICE = 1

const FORMATTED_ACCOUNT_SHORT = '0xf61B...9EEf'
const FORMATTED_BALANCE = '1.123457'
const FORMATTED_PRICE = '$1,000.75'
const FORMATTED_TOKEN_TOTAL = '$1,124.30'
const FORMATTED_WALLET_TOTAL = '$5,621.50'

jest.mock('@usedapp/core', () => ({
  ...jest.requireActual('@usedapp/core'),
  useEtherBalance: jest
    .fn()
    .mockImplementation(() => MOCK_TOKEN_BALANCE_BIG_NUMBER),
  useTokenBalance: jest
    .fn()
    .mockImplementation(() => MOCK_TOKEN_BALANCE_BIG_NUMBER),
}))

jest.mock('../utils/prices', () => ({
  getEthUsdPrice: jest.fn(() => MOCK_TOKEN_USD_PRICE),
  getStethEthPrice: jest.fn(() => MOCK_TOKEN_ETH_PRICE),
  getRethUsdPrice: jest.fn(() => MOCK_TOKEN_USD_PRICE),
  getGrvtUsdPrice: jest.fn(() => MOCK_TOKEN_USD_PRICE),
  getVusdUsdPrice: jest.fn(() => MOCK_TOKEN_USD_PRICE),
}))

const RES_SMALLER_THAN_1280 = false

jest.mock('@chakra-ui/react', () => ({
  __esModule: true,
  ...jest.requireActual('@chakra-ui/react'),
  useMediaQuery: jest.fn(() => [RES_SMALLER_THAN_1280]),
}))

const mockUseMediaQuery = useMediaQuery as jest.Mock<boolean[]>

const renderOptions = {
  wrapperProps: { account: MOCK_ACCOUNT_LONG },
}

describe('Wallet', () => {
  test('Should match snapshot', async () => {
    const { container } = render(
      <Wallet
        name={WALLET_PROVIDER_METAMASK}
        account={MOCK_ACCOUNT_LONG}
        deactivate={jest.fn()}
      />,
      renderOptions
    )

    await screen.findByText(FORMATTED_WALLET_TOTAL)

    expect(container).toMatchSnapshot()
  })

  test('Icon for MetaMask provider should be displayed', async () => {
    render(
      <Wallet
        name={WALLET_PROVIDER_METAMASK}
        account={MOCK_ACCOUNT_LONG}
        deactivate={jest.fn()}
      />,
      renderOptions
    )

    await screen.findByText(FORMATTED_WALLET_TOTAL)

    expect(screen.getByAltText(WALLET_PROVIDER_METAMASK)).toBeInTheDocument()
  })

  test('Icon for Coinbase provider should be displayed', async () => {
    render(
      <Wallet
        name={WALLET_PROVIDER_COINBASE}
        account={MOCK_ACCOUNT_LONG}
        deactivate={jest.fn()}
      />,
      renderOptions
    )

    await screen.findByText(FORMATTED_WALLET_TOTAL)

    expect(screen.getByAltText(WALLET_PROVIDER_COINBASE)).toBeInTheDocument()
  })

  test('Icon for WalletConnect provider should be displayed', async () => {
    render(
      <Wallet
        name={WALLET_PROVIDER_WALLETCONNECT}
        account={MOCK_ACCOUNT_LONG}
        deactivate={jest.fn()}
      />,
      renderOptions
    )

    await screen.findByText(FORMATTED_WALLET_TOTAL)

    expect(
      screen.getByAltText(WALLET_PROVIDER_WALLETCONNECT)
    ).toBeInTheDocument()
  })

  test('Shortened address should be displayed', async () => {
    render(
      <Wallet
        name={WALLET_PROVIDER_METAMASK}
        account={MOCK_ACCOUNT_LONG}
        deactivate={jest.fn()}
      />,
      renderOptions
    )

    await screen.findByText(FORMATTED_WALLET_TOTAL)

    expect(screen.getByText(FORMATTED_ACCOUNT_SHORT)).toBeInTheDocument()
  })

  test('Clicking disconnect should invoke callback', async () => {
    const deactivate = jest.fn()

    render(
      <Wallet
        name={WALLET_PROVIDER_METAMASK}
        account={MOCK_ACCOUNT_LONG}
        deactivate={deactivate}
      />,
      renderOptions
    )

    await screen.findByText(FORMATTED_WALLET_TOTAL)

    const accountButton = screen.getByRole('button', {
      name: /metamask 0xf61B...9EEf/i,
    })

    await userEvent.click(accountButton)

    const disconnectButton = screen.getByRole('menuitem', {
      name: /Disconnect/i,
    })
    await userEvent.click(disconnectButton)

    expect(deactivate).toBeCalled()
  })

  test('Should not render content within accordion', async () => {
    render(
      <Wallet
        name={WALLET_PROVIDER_METAMASK}
        account={MOCK_ACCOUNT_LONG}
        deactivate={jest.fn()}
      />,
      renderOptions
    )

    await screen.findByText(FORMATTED_WALLET_TOTAL)

    const accordionButton = screen.queryByRole('button', {
      name: /Wallet/i,
    })

    expect(accordionButton).not.toBeInTheDocument()
  })

  test('Balances for each token should be displayed and rounded to 6 decimals', async () => {
    render(
      <Wallet
        name={WALLET_PROVIDER_METAMASK}
        account={MOCK_ACCOUNT_LONG}
        deactivate={jest.fn()}
      />,
      renderOptions
    )

    await screen.findByText(FORMATTED_WALLET_TOTAL)

    for (const token in TOKENS) {
      const balanceText = screen.getByTestId(`${token}-balance`)
      within(balanceText).getByText(FORMATTED_BALANCE)
    }
  })

  test('Prices for each token should be displayed', async () => {
    render(
      <Wallet
        name={WALLET_PROVIDER_METAMASK}
        account={MOCK_ACCOUNT_LONG}
        deactivate={jest.fn()}
      />,
      renderOptions
    )

    await screen.findByText(FORMATTED_WALLET_TOTAL)

    for (const token in TOKENS) {
      const balanceText = screen.getByTestId(`${token}-price`)
      within(balanceText).getByText(FORMATTED_PRICE)
    }
  })

  test('Buy link should be rendered for each token', async () => {
    render(
      <Wallet
        name={WALLET_PROVIDER_METAMASK}
        account={MOCK_ACCOUNT_LONG}
        deactivate={jest.fn()}
      />,
      renderOptions
    )

    await screen.findByText(FORMATTED_WALLET_TOTAL)

    for (const token in TOKENS) {
      const buyLink = screen.getByTestId(`${token}-buy`)
      expect(buyLink).toHaveAttribute('href', `https://curve.fi/${token}`)
    }
  })

  test('Totals for each token should be displayed', async () => {
    render(
      <Wallet
        name={WALLET_PROVIDER_METAMASK}
        account={MOCK_ACCOUNT_LONG}
        deactivate={jest.fn()}
      />,
      renderOptions
    )

    await screen.findByText(FORMATTED_WALLET_TOTAL)

    for (const token in TOKENS) {
      const balanceText = screen.getByTestId(`${token}-total`)
      within(balanceText).getByText(FORMATTED_TOKEN_TOTAL)
    }
  })

  test('Wallet total should be displayed', async () => {
    render(
      <Wallet
        name={WALLET_PROVIDER_METAMASK}
        account={MOCK_ACCOUNT_LONG}
        deactivate={jest.fn()}
      />,
      renderOptions
    )

    await screen.findByText(FORMATTED_WALLET_TOTAL)
  })
})

describe('Wallet max width 1280', () => {
  // TODO Figure out how to do snapshots with dynamically generated ids from Chakra UI
  test.skip('Should match snapshot', async () => {
    mockUseMediaQuery.mockReturnValue([true])

    const { container } = render(
      <Wallet
        name={WALLET_PROVIDER_METAMASK}
        account={MOCK_ACCOUNT_LONG}
        deactivate={jest.fn()}
      />,
      renderOptions
    )

    await screen.findByText(FORMATTED_WALLET_TOTAL)

    expect(container).toMatchSnapshot()
  })

  test('Shortened address should be displayed', async () => {
    mockUseMediaQuery.mockReturnValue([true])

    render(
      <Wallet
        name={WALLET_PROVIDER_METAMASK}
        account={MOCK_ACCOUNT_LONG}
        deactivate={jest.fn()}
      />,
      renderOptions
    )

    await screen.findByText(FORMATTED_WALLET_TOTAL)

    expect(screen.getByText(FORMATTED_ACCOUNT_SHORT)).toBeVisible()
  })

  test('Clicking disconnect should invoke callback', async () => {
    mockUseMediaQuery.mockReturnValue([true])
    const deactivate = jest.fn()

    render(
      <Wallet
        name={WALLET_PROVIDER_METAMASK}
        account={MOCK_ACCOUNT_LONG}
        deactivate={deactivate}
      />,
      renderOptions
    )

    await screen.findByText(FORMATTED_WALLET_TOTAL)

    const accountButton = screen.getByRole('button', {
      name: /metamask 0xf61B...9EEf/i,
    })

    await userEvent.click(accountButton)

    const disconnectButton = screen.getByRole('menuitem', {
      name: /Disconnect/i,
    })
    await userEvent.click(disconnectButton)

    expect(deactivate).toBeCalled()
  })

  test('Content should be initially hidden within collapsed accordion', async () => {
    mockUseMediaQuery.mockReturnValue([true])

    render(
      <Wallet
        name={WALLET_PROVIDER_METAMASK}
        account={MOCK_ACCOUNT_LONG}
        deactivate={jest.fn()}
      />,
      renderOptions
    )

    await screen.findByText(FORMATTED_WALLET_TOTAL)

    const accordionButton = screen.getByRole('button', {
      name: /Wallet/i,
    })

    const balanceText = screen.getByText(FORMATTED_WALLET_TOTAL)

    expect(accordionButton).toHaveAttribute('aria-expanded', 'false')
    expect(balanceText).toBeInTheDocument()
    expect(balanceText).not.toBeVisible()
  })

  test('Content should be displayed when accordion is expanded', async () => {
    mockUseMediaQuery.mockReturnValue([true])

    render(
      <Wallet
        name={WALLET_PROVIDER_METAMASK}
        account={MOCK_ACCOUNT_LONG}
        deactivate={jest.fn()}
      />,
      renderOptions
    )

    const accordionButton = screen.getByRole('button', {
      name: /Wallet/i,
    })

    await userEvent.click(accordionButton)
    const balanceText = screen.getByText(FORMATTED_WALLET_TOTAL)

    await waitFor(() => {
      try {
        expect(balanceText).toBeVisible()
        expect(accordionButton).toHaveAttribute('aria-expanded', 'true')
        return Promise.resolve()
      } catch (error) {
        return Promise.reject(error)
      }
    })
  })
})
