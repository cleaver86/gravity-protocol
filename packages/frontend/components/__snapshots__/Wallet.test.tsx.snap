// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wallet Should match snapshot 1`] = `
<div>
  <div
    class="css-1swrp63"
  >
    <div
      class="css-10klw3m"
    >
      <div
        class="css-v57d2b"
      >
        <div
          class="css-10ucka"
        >
          <p
            class="chakra-text css-1u8qc7s"
          >
            Wallet
          </p>
        </div>
        <div
          class="css-17xejub"
        />
        <div
          class="css-s4f5xd"
        >
          <button
            aria-controls="menu-list-:r1:"
            aria-expanded="false"
            aria-haspopup="menu"
            class="chakra-button chakra-menu__menu-button css-z3hoyf"
            id="menu-button-:r1:"
            type="button"
          >
            <span
              class="css-xl71ch"
            >
              <div
                class="css-70qvj9"
              >
                <img
                  alt="MetaMask"
                  class="chakra-image css-xdeki6"
                  src="/images/logo-metamask.png"
                />
                <p
                  class="chakra-text css-mfm9o5"
                >
                  0xf61B...9EEf
                </p>
                <div
                  class="css-1y6ic72"
                >
                  <svg
                    aria-hidden="true"
                    class="svg-inline--fa fa-angle-down  css-4qm0av"
                    data-icon="angle-down"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    viewBox="0 0 448 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M201.4 374.6c12.5 12.5 32.8 12.5 45.3 0l160-160c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L224 306.7 86.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l160 160z"
                      fill="currentColor"
                    />
                  </svg>
                </div>
              </div>
            </span>
          </button>
          <div
            class="css-r6z5ec"
            style="visibility: hidden; position: absolute; min-width: max-content; inset: 0 auto auto 0;"
          >
            <div
              aria-orientation="vertical"
              class="chakra-menu__menu-list css-1vcnbi6"
              id="menu-list-:r1:"
              role="menu"
              style="transform-origin: var(--popper-transform-origin); opacity: 0; visibility: hidden; transform: scale(0.8) translateZ(0);"
              tabindex="-1"
            >
              <button
                class="chakra-menu__menuitem css-h3e82i"
                data-index="0"
                id="menu-list-:r1:-menuitem-:r2:"
                role="menuitem"
                tabindex="-1"
                type="button"
              >
                Disconnect
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="css-cf1nu1"
      >
        <div
          class="css-1sb7ptk"
        >
          <div
            class="css-tcv3m"
          >
            <div
              class="chakra-stack css-zsoya5"
            >
              <div
                class="css-k008qs"
              >
                <div
                  class="css-1wjoefq"
                >
                  <div
                    class="css-1nsz27u"
                  >
                    <svg />
                  </div>
                </div>
                <div
                  class="css-hboir5"
                >
                  <div
                    class="css-0"
                  >
                    <div
                      class="css-3gj7jk"
                    >
                      <p
                        class="chakra-text css-s8307y"
                        data-testid="grvt-balance"
                      >
                        1.123457
                      </p>
                      <p
                        class="chakra-text css-1qdnhws"
                      >
                        GRVT
                      </p>
                    </div>
                    <p
                      class="chakra-text css-1icp0bk"
                      data-testid="grvt-total"
                    >
                      $1,124.30
                    </p>
                  </div>
                  <div
                    class="css-17xejub"
                  />
                  <div
                    class="css-s2uf1z"
                  >
                    <p
                      class="chakra-text css-1icp0bk"
                      data-testid="grvt-price"
                    >
                      $1,000.75
                    </p>
                    <a
                      class="chakra-link css-1h52md4"
                      data-testid="grvt-buy"
                      href="https://curve.fi/grvt"
                      role="link"
                    >
                      Buy
                    </a>
                  </div>
                </div>
              </div>
              <div
                class="chakra-stack__divider css-l0hi3m"
              />
              <div
                class="css-k008qs"
              >
                <div
                  class="css-1wjoefq"
                >
                  <div
                    class="css-1nsz27u"
                  >
                    <svg />
                  </div>
                </div>
                <div
                  class="css-hboir5"
                >
                  <div
                    class="css-0"
                  >
                    <div
                      class="css-3gj7jk"
                    >
                      <p
                        class="chakra-text css-s8307y"
                        data-testid="vusd-balance"
                      >
                        1.123457
                      </p>
                      <p
                        class="chakra-text css-1qdnhws"
                      >
                        VUSD
                      </p>
                    </div>
                    <p
                      class="chakra-text css-1icp0bk"
                      data-testid="vusd-total"
                    >
                      $1,124.30
                    </p>
                  </div>
                  <div
                    class="css-17xejub"
                  />
                  <div
                    class="css-s2uf1z"
                  >
                    <p
                      class="chakra-text css-1icp0bk"
                      data-testid="vusd-price"
                    >
                      $1,000.75
                    </p>
                    <a
                      class="chakra-link css-1h52md4"
                      data-testid="vusd-buy"
                      href="https://curve.fi/vusd"
                      role="link"
                    >
                      Buy
                    </a>
                  </div>
                </div>
              </div>
              <div
                class="chakra-stack__divider css-l0hi3m"
              />
              <div
                class="css-k008qs"
              >
                <div
                  class="css-1wjoefq"
                >
                  <div
                    class="css-1nsz27u"
                  >
                    <svg />
                  </div>
                </div>
                <div
                  class="css-hboir5"
                >
                  <div
                    class="css-0"
                  >
                    <div
                      class="css-3gj7jk"
                    >
                      <p
                        class="chakra-text css-s8307y"
                        data-testid="eth-balance"
                      >
                        1.123457
                      </p>
                      <p
                        class="chakra-text css-1qdnhws"
                      >
                        ETH
                      </p>
                    </div>
                    <p
                      class="chakra-text css-1icp0bk"
                      data-testid="eth-total"
                    >
                      $1,124.30
                    </p>
                  </div>
                  <div
                    class="css-17xejub"
                  />
                  <div
                    class="css-s2uf1z"
                  >
                    <p
                      class="chakra-text css-1icp0bk"
                      data-testid="eth-price"
                    >
                      $1,000.75
                    </p>
                    <a
                      class="chakra-link css-1h52md4"
                      data-testid="eth-buy"
                      href="https://curve.fi/eth"
                      role="link"
                    >
                      Buy
                    </a>
                  </div>
                </div>
              </div>
              <div
                class="chakra-stack__divider css-l0hi3m"
              />
              <div
                class="css-k008qs"
              >
                <div
                  class="css-1wjoefq"
                >
                  <div
                    class="css-1nsz27u"
                  >
                    <svg />
                  </div>
                </div>
                <div
                  class="css-hboir5"
                >
                  <div
                    class="css-0"
                  >
                    <div
                      class="css-3gj7jk"
                    >
                      <p
                        class="chakra-text css-s8307y"
                        data-testid="reth-balance"
                      >
                        1.123457
                      </p>
                      <p
                        class="chakra-text css-1qdnhws"
                      >
                        RETH
                      </p>
                    </div>
                    <p
                      class="chakra-text css-1icp0bk"
                      data-testid="reth-total"
                    >
                      $1,124.30
                    </p>
                  </div>
                  <div
                    class="css-17xejub"
                  />
                  <div
                    class="css-s2uf1z"
                  >
                    <p
                      class="chakra-text css-1icp0bk"
                      data-testid="reth-price"
                    >
                      $1,000.75
                    </p>
                    <a
                      class="chakra-link css-1h52md4"
                      data-testid="reth-buy"
                      href="https://curve.fi/reth"
                      role="link"
                    >
                      Buy
                    </a>
                  </div>
                </div>
              </div>
              <div
                class="chakra-stack__divider css-l0hi3m"
              />
              <div
                class="css-k008qs"
              >
                <div
                  class="css-1wjoefq"
                >
                  <div
                    class="css-1nsz27u"
                  >
                    <svg />
                  </div>
                </div>
                <div
                  class="css-hboir5"
                >
                  <div
                    class="css-0"
                  >
                    <div
                      class="css-3gj7jk"
                    >
                      <p
                        class="chakra-text css-s8307y"
                        data-testid="steth-balance"
                      >
                        1.123457
                      </p>
                      <p
                        class="chakra-text css-1qdnhws"
                      >
                        STETH
                      </p>
                    </div>
                    <p
                      class="chakra-text css-1icp0bk"
                      data-testid="steth-total"
                    >
                      $1,124.30
                    </p>
                  </div>
                  <div
                    class="css-17xejub"
                  />
                  <div
                    class="css-s2uf1z"
                  >
                    <p
                      class="chakra-text css-1icp0bk"
                      data-testid="steth-price"
                    >
                      $1,000.75
                    </p>
                    <a
                      class="chakra-link css-1h52md4"
                      data-testid="steth-buy"
                      href="https://curve.fi/steth"
                      role="link"
                    >
                      Buy
                    </a>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="css-17xejub"
          />
          <div
            class="css-xrdnfo"
          >
            <div
              class="css-1rwzck"
            >
              <p
                class="chakra-text css-1u8qc7s"
              >
                Total
              </p>
            </div>
            <p
              class="chakra-text css-1vxklou"
            >
              $5,621.50
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;

exports[`Wallet max width 1280 Should match snapshot 1`] = `
<div>
  <div
    class="css-1aw2xze"
  >
    <div
      class="chakra-accordion css-1whkjwr"
    >
      <div
        class="chakra-accordion__item css-1t7rcca"
      >
        <div
          class="css-13luk2y"
        >
          <div
            class="css-s4f5xd"
          >
            <button
              aria-controls="menu-list-:r2:"
              aria-expanded="false"
              aria-haspopup="menu"
              class="chakra-button chakra-menu__menu-button css-z3hoyf"
              id="menu-button-:r2:"
              type="button"
            >
              <span
                class="css-xl71ch"
              >
                <div
                  class="css-70qvj9"
                >
                  <img
                    alt="MetaMask"
                    class="chakra-image css-xdeki6"
                    src="/images/logo-metamask.png"
                  />
                  <p
                    class="chakra-text css-mfm9o5"
                  >
                    0xf61B...9EEf
                  </p>
                  <div
                    class="css-1y6ic72"
                  >
                    <svg
                      aria-hidden="true"
                      class="svg-inline--fa fa-angle-down  css-4qm0av"
                      data-icon="angle-down"
                      data-prefix="fas"
                      focusable="false"
                      role="img"
                      viewBox="0 0 448 512"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M201.4 374.6c12.5 12.5 32.8 12.5 45.3 0l160-160c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L224 306.7 86.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l160 160z"
                        fill="currentColor"
                      />
                    </svg>
                  </div>
                </div>
              </span>
            </button>
            <div
              class="css-r6z5ec"
              style="visibility: hidden; position: absolute; min-width: max-content; inset: 0 auto auto 0;"
            >
              <div
                aria-orientation="vertical"
                class="chakra-menu__menu-list css-1vcnbi6"
                id="menu-list-:r2:"
                role="menu"
                style="transform-origin: var(--popper-transform-origin); opacity: 0; visibility: hidden; transform: scale(0.8) translateZ(0);"
                tabindex="-1"
              >
                <button
                  class="chakra-menu__menuitem css-h3e82i"
                  data-index="0"
                  id="menu-list-:r2:-menuitem-:r3:"
                  role="menuitem"
                  tabindex="-1"
                  type="button"
                >
                  Disconnect
                </button>
              </div>
            </div>
          </div>
        </div>
        <button
          aria-controls="accordion-panel-:r0:"
          aria-expanded="false"
          class="chakra-accordion__button css-a5spwg"
          data-index="0"
          id="accordion-button-:r0:"
          type="button"
        >
          <div
            class="css-k008qs"
          >
            <svg
              aria-hidden="true"
              class="chakra-icon chakra-accordion__icon css-186l2rg"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                fill="currentColor"
              />
            </svg>
            <div
              class="css-10ucka"
            >
              <p
                class="chakra-text css-1u8qc7s"
              >
                Wallet
              </p>
            </div>
          </div>
        </button>
        <div
          class="chakra-collapse"
          style="overflow: hidden; display: block; opacity: 0; height: auto;"
        >
          <div
            aria-labelledby="accordion-button-:r0:"
            class="chakra-accordion__panel css-gja1qb"
            id="accordion-panel-:r0:"
            role="region"
          >
            <div
              class="css-1sb7ptk"
            >
              <div
                class="css-tcv3m"
              >
                <div
                  class="chakra-stack css-zsoya5"
                >
                  <div
                    class="css-k008qs"
                  >
                    <div
                      class="css-1wjoefq"
                    >
                      <div
                        class="css-1nsz27u"
                      >
                        <svg />
                      </div>
                    </div>
                    <div
                      class="css-hboir5"
                    >
                      <div
                        class="css-0"
                      >
                        <div
                          class="css-3gj7jk"
                        >
                          <p
                            class="chakra-text css-s8307y"
                            data-testid="grvt-balance"
                          >
                            1.123457
                          </p>
                          <p
                            class="chakra-text css-1qdnhws"
                          >
                            GRVT
                          </p>
                        </div>
                        <p
                          class="chakra-text css-1icp0bk"
                          data-testid="grvt-total"
                        >
                          $1,124.30
                        </p>
                      </div>
                      <div
                        class="css-17xejub"
                      />
                      <div
                        class="css-s2uf1z"
                      >
                        <p
                          class="chakra-text css-1icp0bk"
                          data-testid="grvt-price"
                        >
                          $1,000.75
                        </p>
                        <a
                          class="chakra-link css-1h52md4"
                          href="https://curve.fi/grvt"
                        >
                          Buy
                        </a>
                      </div>
                    </div>
                  </div>
                  <div
                    class="chakra-stack__divider css-l0hi3m"
                  />
                  <div
                    class="css-k008qs"
                  >
                    <div
                      class="css-1wjoefq"
                    >
                      <div
                        class="css-1nsz27u"
                      >
                        <svg />
                      </div>
                    </div>
                    <div
                      class="css-hboir5"
                    >
                      <div
                        class="css-0"
                      >
                        <div
                          class="css-3gj7jk"
                        >
                          <p
                            class="chakra-text css-s8307y"
                            data-testid="vusd-balance"
                          >
                            1.123457
                          </p>
                          <p
                            class="chakra-text css-1qdnhws"
                          >
                            VUSD
                          </p>
                        </div>
                        <p
                          class="chakra-text css-1icp0bk"
                          data-testid="vusd-total"
                        >
                          $1,124.30
                        </p>
                      </div>
                      <div
                        class="css-17xejub"
                      />
                      <div
                        class="css-s2uf1z"
                      >
                        <p
                          class="chakra-text css-1icp0bk"
                          data-testid="vusd-price"
                        >
                          $1,000.75
                        </p>
                        <a
                          class="chakra-link css-1h52md4"
                          href="https://curve.fi/vusd"
                        >
                          Buy
                        </a>
                      </div>
                    </div>
                  </div>
                  <div
                    class="chakra-stack__divider css-l0hi3m"
                  />
                  <div
                    class="css-k008qs"
                  >
                    <div
                      class="css-1wjoefq"
                    >
                      <div
                        class="css-1nsz27u"
                      >
                        <svg />
                      </div>
                    </div>
                    <div
                      class="css-hboir5"
                    >
                      <div
                        class="css-0"
                      >
                        <div
                          class="css-3gj7jk"
                        >
                          <p
                            class="chakra-text css-s8307y"
                            data-testid="eth-balance"
                          >
                            1.123457
                          </p>
                          <p
                            class="chakra-text css-1qdnhws"
                          >
                            ETH
                          </p>
                        </div>
                        <p
                          class="chakra-text css-1icp0bk"
                          data-testid="eth-total"
                        >
                          $1,124.30
                        </p>
                      </div>
                      <div
                        class="css-17xejub"
                      />
                      <div
                        class="css-s2uf1z"
                      >
                        <p
                          class="chakra-text css-1icp0bk"
                          data-testid="eth-price"
                        >
                          $1,000.75
                        </p>
                        <a
                          class="chakra-link css-1h52md4"
                          href="https://curve.fi/eth"
                        >
                          Buy
                        </a>
                      </div>
                    </div>
                  </div>
                  <div
                    class="chakra-stack__divider css-l0hi3m"
                  />
                  <div
                    class="css-k008qs"
                  >
                    <div
                      class="css-1wjoefq"
                    >
                      <div
                        class="css-1nsz27u"
                      >
                        <svg />
                      </div>
                    </div>
                    <div
                      class="css-hboir5"
                    >
                      <div
                        class="css-0"
                      >
                        <div
                          class="css-3gj7jk"
                        >
                          <p
                            class="chakra-text css-s8307y"
                            data-testid="reth-balance"
                          >
                            1.123457
                          </p>
                          <p
                            class="chakra-text css-1qdnhws"
                          >
                            rETH
                          </p>
                        </div>
                        <p
                          class="chakra-text css-1icp0bk"
                          data-testid="reth-total"
                        >
                          $1,124.30
                        </p>
                      </div>
                      <div
                        class="css-17xejub"
                      />
                      <div
                        class="css-s2uf1z"
                      >
                        <p
                          class="chakra-text css-1icp0bk"
                          data-testid="reth-price"
                        >
                          $1,000.75
                        </p>
                        <a
                          class="chakra-link css-1h52md4"
                          href="https://curve.fi/reth"
                        >
                          Buy
                        </a>
                      </div>
                    </div>
                  </div>
                  <div
                    class="chakra-stack__divider css-l0hi3m"
                  />
                  <div
                    class="css-k008qs"
                  >
                    <div
                      class="css-1wjoefq"
                    >
                      <div
                        class="css-1nsz27u"
                      >
                        <svg />
                      </div>
                    </div>
                    <div
                      class="css-hboir5"
                    >
                      <div
                        class="css-0"
                      >
                        <div
                          class="css-3gj7jk"
                        >
                          <p
                            class="chakra-text css-s8307y"
                            data-testid="steth-balance"
                          >
                            1.123457
                          </p>
                          <p
                            class="chakra-text css-1qdnhws"
                          >
                            stETH
                          </p>
                        </div>
                        <p
                          class="chakra-text css-1icp0bk"
                          data-testid="steth-total"
                        >
                          $1,124.30
                        </p>
                      </div>
                      <div
                        class="css-17xejub"
                      />
                      <div
                        class="css-s2uf1z"
                      >
                        <p
                          class="chakra-text css-1icp0bk"
                          data-testid="steth-price"
                        >
                          $1,000.75
                        </p>
                        <a
                          class="chakra-link css-1h52md4"
                          href="https://curve.fi/steth"
                        >
                          Buy
                        </a>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="css-17xejub"
              />
              <div
                class="css-xrdnfo"
              >
                <div
                  class="css-1rwzck"
                >
                  <p
                    class="chakra-text css-1u8qc7s"
                  >
                    Total
                  </p>
                </div>
                <p
                  class="chakra-text css-1vxklou"
                >
                  $5,621.50
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
